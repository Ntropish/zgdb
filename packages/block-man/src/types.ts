export type Address = Uint8Array;

export interface BlockStore {
  /**
   * Retrieves a block of data from the store.
   * @param address The content address of the block.
   * @returns A Promise that resolves to the block's data as a Uint8Array, or undefined if not found.
   */
  get: (address: Address) => Promise<Uint8Array | undefined>;

  /**
   * Stores a block of data.
   * @param block The block data to store.
   * @returns A Promise that resolves to the content address of the stored block.
   */
  put: (block: Uint8Array) => Promise<Address>;

  /**
   * Checks for the existence of a block.
   * @param address The content address of the block.
   * @returns A Promise that resolves to true if the block exists, false otherwise.
   */
  has: (address: Address) => Promise<boolean>;
}

export interface Hasher {
  (data: Uint8Array): Promise<Address>;
}
