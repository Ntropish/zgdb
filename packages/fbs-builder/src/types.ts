export type FbsScalar =
  | "bool"
  | "byte"
  | "ubyte"
  | "short"
  | "ushort"
  | "int"
  | "uint"
  | "long"
  | "ulong"
  | "float"
  | "double";

export type FbsType = FbsScalar | string; // string for other tables, structs, etc.

export interface FbsField {
  name: string;
  type: FbsType;
  defaultValue?: number | string | boolean | null;
  isVector: boolean;
  attributes: Map<string, string | boolean>;
}

export interface FbsDeclaration {
  kind: "table" | "struct" | "enum" | "union";
  name: string;
  docs: string[];
}

export interface FbsTableState extends FbsDeclaration {
  kind: "table";
  fields: FbsField[];
  attributes: Map<string, string | boolean>;
}

export interface FbsStructState extends FbsDeclaration {
  kind: "struct";
  fields: FbsField[]; // in structs, fields cannot be optional or have defaults
}

export interface FbsEnumState extends FbsDeclaration {
  kind: "enum";
  type: FbsScalar;
  values: Map<string, number>;
}

export interface FbsUnionState extends FbsDeclaration {
  kind: "union";
  values: string[]; // table names
}

export interface FbsFileState {
  namespace: string | null;
  includes: string[];
  declarations: FbsDeclaration[];
  rootType: string | null;
  fileIdentifier: string | null;
  fileExtension: string | null;
}
