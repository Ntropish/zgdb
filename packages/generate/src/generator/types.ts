import { NormalizedSchema } from "../parser/types.js";

export interface GeneratorConfig {
  schemas: NormalizedSchema[];
  outputDirectory: string;
  options: {
    /**
     * The file extension to use for the generated import of the FlatBuffers schema.
     * E.g., '.js', '.mjs', or an empty string.
     * @default ".js"
     */
    importExtension?: string;
  };
}
