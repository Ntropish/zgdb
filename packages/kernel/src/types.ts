import { ComponentFactory } from "./component";

/**
 * A unique symbol used to signal that an orchestrator pipeline should halt execution.
 */
export const BREAK = Symbol.for("Orchestrator.BREAK");

/**
 * The most basic execution unit in the kernel. A function that accepts a context object
 * and can perform synchronous or asynchronous work.
 */
export type StepHandler<TContext extends Record<string, any> = any> = (
  ctx: TContext
) => unknown;

/**
 * The interface for a kernel that executes a linear pipeline of StepHandlers.
 */
export interface OrchestratorKernel<
  TContext extends Record<string, any> = any
> {
  run(ctx?: Partial<TContext>): Promise<TContext | typeof BREAK>;
  clone(): OrchestratorKernel<TContext>;
}

/**
 * A reactor's event handler is a self-contained orchestrator pipeline.
 */
export type EventHandler<TContext extends Record<string, any> = any> =
  OrchestratorKernel<TContext>;

/**
 * A structured collection of StepHandler arrays for logging at different levels.
 * This is designed to be injected into kernels to provide logging capabilities
 * without creating circular dependencies.
 */
export type LoggerPlugins = {
  info?: StepHandler[];
  warn?: StepHandler[];
  error?: StepHandler[];
};
