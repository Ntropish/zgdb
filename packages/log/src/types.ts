export type LogLevel = "fatal" | "error" | "warn" | "info" | "debug" | "trace";

export const LogLevels: Record<LogLevel, number> = {
  fatal: 60,
  error: 50,
  warn: 40,
  info: 30,
  debug: 20,
  trace: 10,
};

export interface LogRecord {
  level: LogLevel;
  message: string;
  time: number;
  data?: Record<string, unknown>;
  error?: Error;
}

export interface Transport {
  send(payload: string): void | Promise<void>;
}
