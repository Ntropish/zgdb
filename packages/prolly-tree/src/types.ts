export type Diff = {
  key: Uint8Array;
  localValue?: Uint8Array;
  remoteValue?: Uint8Array;
};

export type ConflictResolver = (
  key: Uint8Array,
  base: Uint8Array | undefined,
  local: Uint8Array,
  remote: Uint8Array
) => Promise<Uint8Array>;
