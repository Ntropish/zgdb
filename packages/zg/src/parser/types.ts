/**
 * This file defines the Intermediate Representation (IR) for our schema,
 * as well as the shape of the raw, user-defined schema files.
 */

import { z } from "zod";

// --- User-Facing Types ---

export type IndexDef = {
  on: string | string[];
  type?: "btree" | "hash" | "fulltext";
  unique?: boolean;
  description?: string;
};

export type PolymorphicRelationshipDef = {
  type: "polymorphic";
  cardinality: "one" | "many";
  required?: boolean;
  description?: string;
  discriminator: string; // e.g., 'targetType'
  foreignKey: string; // e.g., 'targetId'
  references: string[]; // e.g., ['Post', 'Comment']
};

export type StandardRelationshipDef = {
  type: "standard";
  entity: string;
  field?: string;
  cardinality: "one" | "many";
  required?: boolean;
  description?: string;
  mappedBy?: string;
};

export type RelationshipDef =
  | StandardRelationshipDef
  | PolymorphicRelationshipDef;

export type ManyToManyDef = {
  node: string;
  through: string;
  myKey: string;
  theirKey: string;
  description?: string;
};

export type EntityDef = {
  name: string;
  description?: string;
  schema: z.ZodObject<any>;
  relationships?: Record<string, RelationshipDef>;
  manyToMany?: Record<string, ManyToManyDef>;
  indexes?: IndexDef[];
  resolvers?: Record<string, Resolver<any>>;
};

export type ResolverContext<TClient = any, TActor = any> = {
  actor: TActor;
  db: TClient;
  input?: Record<string, any>;
  record?: Record<string, any>;
};

export type Resolver<TClient = any, TActor = any> = (
  context: ResolverContext<TClient, TActor>
) => Promise<boolean> | boolean;

export type Policy<TClient = any, TActor = any> = Resolver<TClient, TActor>;

export type InferredResolvers<
  TClient,
  TActor,
  TEntities extends Record<string, EntityDef>
> = {
  [K in keyof TEntities]?: Record<string, Resolver<TClient, TActor>>;
};

export type AuthBlock = {
  create?: string | string[];
  read?: string | string[];
  update?: string | string[];
  delete?: string | string[];
};

export type SchemaConfig<
  TClient = any,
  TActor = any,
  TEntities extends Record<string, EntityDef> = any,
  TGlobalResolvers extends Record<string, Resolver<TClient, TActor>> = any,
  TResolvers extends InferredResolvers<TClient, TActor, TEntities> = any
> = {
  entities: TEntities;
  globalResolvers?: TGlobalResolvers;
  resolvers?: TResolvers;
  auth?: {
    [entityName: string]: AuthBlock;
  };
};

// --- Internal, Normalized Types ---

export type Field = {
  name: string;
  type: string;
  attributes: Map<string, string | boolean>;
  description?: string;
  isVector?: boolean;
  required: boolean;
};

export type Index = {
  name: string;
  on: string[];
  unique?: boolean;
  type: "btree" | "hash" | "fulltext";
};

export type Relationship = {
  name: string;
  node: string;
  cardinality: "one" | "many";
  required?: boolean;
  description?: string;
  mappedBy?: string;
};

export type ManyToManyRelationship = {
  name: string;
  node: string;
  through: string;
  myKey: string;
  theirKey: string;
  description?: string;
};

export type PolymorphicRelationship = {
  type: "polymorphic";
  field: string;
};

export type NormalizedSchema = {
  name: string;
  description?: string;
  fields: Field[];
  isJoinTable?: boolean;
  indexes: Index[];
  auth?: AuthBlock;
  localResolvers?: Record<string, Function>;
  globalResolvers?: Record<string, Function>;
  relationships: (Relationship | PolymorphicRelationship)[];
  manyToMany: ManyToManyRelationship[];
};
