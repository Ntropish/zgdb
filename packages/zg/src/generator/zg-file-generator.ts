import {
  NormalizedSchema,
  Field,
  Relationship,
  ManyToManyRelationship,
} from "../parser/types.js";
import { topologicalSort } from "./topological-sort.js";

function generateTsField(field: Field): string {
  // This needs to map FBS types to TS types
  const typeMap: Record<string, string> = {
    string: "string",
    long: "number",
    bool: "boolean",
  };

  // Handle table types, e.g., User_Profile -> Promise<ZGUserProfile>
  const isTableType = !typeMap[field.type] && !field.type.startsWith("[");
  const isTableArrayType = field.type.startsWith("[");

  let tsType = "any";
  if (typeMap[field.type]) {
    tsType = typeMap[field.type];
  } else if (isTableType) {
    tsType = `Promise<ZG${field.type} | null>`;
  } else if (isTableArrayType) {
    const baseType = field.type.replace(/\[|\]/g, "");
    tsType = `Promise<ZG${baseType}[] | null>`;
  }

  return `  ${field.name}: ${tsType};`;
}

function generateTsInterface(schema: NormalizedSchema): string {
  const fields = schema.fields.map(generateTsField).join("\n");

  const rels = schema.relationships
    .map((rel) => {
      const nodeType = `ZG${rel.node}`;
      const type =
        rel.cardinality === "one"
          ? `Promise<${nodeType} | null>`
          : `Promise<${nodeType}[]>`;
      return `  ${rel.name}: ${type};`;
    })
    .join("\n");

  const m2mRels = schema.manyToMany
    .map((rel) => {
      const nodeType = `ZG${rel.node}`;
      return `  ${rel.name}: Promise<${nodeType}[]>;`;
    })
    .join("\n");

  const allFields = [fields, rels, m2mRels].filter(Boolean).join("\n\n");

  return `export interface ZG${schema.name} {\n${allFields}\n}`;
}

function generateMetaData(schema: NormalizedSchema): string {
  const rels = schema.relationships
    .map(
      (rel) =>
        `    ${rel.name}: { type: '${rel.cardinality}', node: '${rel.node}'${
          rel.mappedBy ? `, mappedBy: '${rel.mappedBy}'` : ""
        } },`
    )
    .join("\n");

  const m2mRels = schema.manyToMany
    .map(
      (rel) =>
        `    ${rel.name}: { type: 'many', node: '${rel.node}', through: '${rel.through}' },`
    )
    .join("\n");

  return `export const ${schema.name}Meta = {
  getRootAs: LowLevel.${schema.name}.getRootAs${schema.name},
  relationships: {
${rels}
  },
  manyToMany: {
${m2mRels}
  },
};`;
}

/**
 * Generates the content for the high-level schema.zg.ts file.
 * @param schemas An array of all normalized schemas.
 * @returns The content of the schema.zg.ts file as a string.
 */
export function generateZgFile(schemas: NormalizedSchema[]): string {
  const fileHeader = `// Generated by ZG. Do not edit.\n\n`;
  const imports = `import { ZGDatabase, ZGNode } from '@tsmk/zg-client';\nimport * as LowLevel from './schema_generated';\n\n`;

  // Sort schemas to ensure consistent output order
  const sortedSchemas = topologicalSort(schemas);

  const interfaces = sortedSchemas.map(generateTsInterface).join("\n\n");
  const metadata = sortedSchemas.map(generateMetaData).join("\n\n");

  const schemaManifest = `export const schema = {\n${sortedSchemas
    .map((s) => `  ${s.name}: ${s.name}Meta,`)
    .join("\n")}\n};`;

  const dbInterfaceEntries = sortedSchemas
    .map((s) => `  ${s.name}: ZG${s.name};`)
    .join("\n");
  const createDbFn = `\nexport function createDatabase(config: any): ZGDatabase<{\n${dbInterfaceEntries}\n}> {\n  return new ZGDatabase(schema, config);\n}`;

  return (
    fileHeader +
    imports +
    interfaces +
    "\n\n" +
    metadata +
    "\n\n" +
    schemaManifest +
    "\n" +
    createDbFn +
    "\n"
  );
}
