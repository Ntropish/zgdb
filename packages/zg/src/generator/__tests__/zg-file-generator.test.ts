import { describe, it, expect } from "vitest";
import { ZgFileGenerator } from "../zg-file-generator.js";
import { NormalizedSchema } from "../../parser/types.js";

const testSchema: NormalizedSchema[] = [
  {
    name: "User",
    fields: [
      { name: "id", type: "ID", required: true, attributes: new Map() },
      { name: "name", type: "string", required: true, attributes: new Map() },
    ],
    indexes: [],
    relationships: [
      {
        name: "posts",
        node: "Post",
        cardinality: "many",
        mappedBy: "author",
      },
    ],
    manyToMany: [],
  },
  {
    name: "Post",
    fields: [
      { name: "id", type: "ID", required: true, attributes: new Map() },
      { name: "title", type: "string", required: true, attributes: new Map() },
      { name: "authorId", type: "ID", required: true, attributes: new Map() },
    ],
    indexes: [],
    relationships: [
      {
        name: "author",
        node: "User",
        cardinality: "one",
        foreignKey: "authorId",
      },
    ],
    manyToMany: [],
  },
];

describe("ZgFileGenerator", () => {
  it("should generate a complete schema file from all sub-generators", () => {
    const generator = new ZgFileGenerator();
    const result = generator.generate(testSchema);
    expect(result).toMatchSnapshot();
  });
});
