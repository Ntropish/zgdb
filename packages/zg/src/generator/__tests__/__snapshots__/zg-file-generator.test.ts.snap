// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ZgFileGenerator > should generate a complete schema file from all sub-generators 1`] = `
"// @generated
// Automatically generated. Don't change this file manually.
// Name: schema.ts


// @ts-nocheck
// This file is generated by @zgdb/zg. Do not edit it manually.
// TODO: Generate based on the user's config
import { ZgDatabase, ZgBaseNode, ZgAuthContext, EntityCollection } from '@zgdb/client';
import { Builder, ByteBuffer } from 'flatbuffers';
import * as UserFB from './schema/user.js';
import * as PostFB from './schema/post.js';


// --- Interfaces ---
export interface IUser {
  id: any;
  name: string;
}

export interface IPost {
  id: any;
  title: string;
  authorId: any;
}

// --- Create Input Types ---
export type CreateUserInput = { name: string };
export type CreatePostInput = { title: string, authorId: any };

// --- Node Classes ---
export class UserNode<TActor> extends ZgBaseNode<UserFB.User, TActor> {
  constructor(
    db: ZgDatabase,
    fbb: UserFB.User,
    authContext: ZgAuthContext<TActor> | null
  ) {
    super(db, fbb, authContext);
    
    return new Proxy(this, {
      get: (target, prop, receiver) => {
        const entityResolvers = target.db.config.entityResolvers?.['User'] ?? {};
        if (prop in entityResolvers) {
          return entityResolvers[prop as keyof typeof entityResolvers]({ actor: target.authContext?.actor, db: target.db, node: target });
        }
        const globalResolvers = target.db.config.globalResolvers ?? {};
        if (prop in globalResolvers) {
          return globalResolvers[prop as keyof typeof globalResolvers]({ actor: target.authContext?.actor, db: target.db, node: target });
        }
        return Reflect.get(target, prop, receiver);
      },
      set: (target, prop, value, receiver) => {
        const schemaFields = new Set(['id', 'name']);
        if (!schemaFields.has(prop as string)) {
          return Reflect.set(target, prop, value, receiver);
        }

        const builder = new Builder(1024);
        const data = { id: (prop === 'id') ? value : target.fbb.id(), name: (prop === 'name') ? value : target.fbb.name() };
        
    const nameOffset = data.name ? builder.createString(data.name) : 0;
        
        const entityOffset = UserFB.User.createUser(builder, data.id, nameOffset);
        builder.finish(entityOffset);
        const buffer = builder.asUint8Array();

        target.db.update(
          'User',
          target.id,
          buffer
        );
        
        const newFbb = UserFB.User.getRootAsUser(new ByteBuffer(buffer));
        target.fbb = newFbb;

        return true;
      }
    });
  }

  // --- Fields ---
  get id(): any {
    return this.fbb.id();
  }

  get name(): string | null {
    return this.fbb.name();
  }

  // --- Relationships ---

}

export class PostNode<TActor> extends ZgBaseNode<PostFB.Post, TActor> {
  constructor(
    db: ZgDatabase,
    fbb: PostFB.Post,
    authContext: ZgAuthContext<TActor> | null
  ) {
    super(db, fbb, authContext);
    
    return new Proxy(this, {
      get: (target, prop, receiver) => {
        const entityResolvers = target.db.config.entityResolvers?.['Post'] ?? {};
        if (prop in entityResolvers) {
          return entityResolvers[prop as keyof typeof entityResolvers]({ actor: target.authContext?.actor, db: target.db, node: target });
        }
        const globalResolvers = target.db.config.globalResolvers ?? {};
        if (prop in globalResolvers) {
          return globalResolvers[prop as keyof typeof globalResolvers]({ actor: target.authContext?.actor, db: target.db, node: target });
        }
        return Reflect.get(target, prop, receiver);
      },
      set: (target, prop, value, receiver) => {
        const schemaFields = new Set(['id', 'title', 'authorId']);
        if (!schemaFields.has(prop as string)) {
          return Reflect.set(target, prop, value, receiver);
        }

        const builder = new Builder(1024);
        const data = { authorId: (prop === 'authorId') ? value : target.fbb.authorId(), id: (prop === 'id') ? value : target.fbb.id(), title: (prop === 'title') ? value : target.fbb.title() };
        
    const titleOffset = data.title ? builder.createString(data.title) : 0;
        
        const entityOffset = PostFB.Post.createPost(builder, data.authorId, data.id, titleOffset);
        builder.finish(entityOffset);
        const buffer = builder.asUint8Array();

        target.db.update(
          'Post',
          target.id,
          buffer
        );
        
        const newFbb = PostFB.Post.getRootAsPost(new ByteBuffer(buffer));
        target.fbb = newFbb;

        return true;
      }
    });
  }

  // --- Fields ---
  get id(): any {
    return this.fbb.id();
  }

  get title(): string | null {
    return this.fbb.title();
  }

  get authorId(): any {
    return this.fbb.authorId();
  }

  // --- Relationships ---

}

// --- Collection Classes ---
export class UserCollection<TActor> {
    constructor(
      private db: ZgDatabase,
      private authContext: ZgAuthContext<TActor> | null
    ) {}

    get collectionName(): string {
      return 'User';
    }

    create(data: CreateUserInput): UserNode<TActor> {
      const builder = new Builder(1024);
      
          const nameOffset = data.name ? builder.createString(data.name) : 0;
      

      UserFB.User.startUser(builder);
      
      UserFB.User.addId(builder, data.id ?? null);
    UserFB.User.addName(builder, nameOffset ?? 0);

      const entityOffset = UserFB.User.endUser(builder);
      
      builder.finish(entityOffset);
      const buffer = builder.asUint8Array();

      this.db.insert(this.collectionName, data.id, buffer);

      const fbb = UserFB.User.getRootAsUser(new ByteBuffer(buffer));

      return new UserNode<TActor>(
        this.db,
        fbb,
        this.authContext,
      );
    }

    *[Symbol.iterator](): Generator<UserNode<TActor>> {
      for (const [id, buffer] of this.db.scan(this.collectionName)) {
        const fbb = UserFB.User.getRootAsUser(new ByteBuffer(buffer));
        yield new UserNode<TActor>(
          this.db,
          fbb,
          this.authContext,
        );
      }
    }
  }

export class PostCollection<TActor> {
    constructor(
      private db: ZgDatabase,
      private authContext: ZgAuthContext<TActor> | null
    ) {}

    get collectionName(): string {
      return 'Post';
    }

    create(data: CreatePostInput): PostNode<TActor> {
      const builder = new Builder(1024);
      
          const titleOffset = data.title ? builder.createString(data.title) : 0;
      

      PostFB.Post.startPost(builder);
      
      PostFB.Post.addAuthorId(builder, data.authorId ?? null);
    PostFB.Post.addId(builder, data.id ?? null);
    PostFB.Post.addTitle(builder, titleOffset ?? 0);

      const entityOffset = PostFB.Post.endPost(builder);
      
      builder.finish(entityOffset);
      const buffer = builder.asUint8Array();

      this.db.insert(this.collectionName, data.id, buffer);

      const fbb = PostFB.Post.getRootAsPost(new ByteBuffer(buffer));

      return new PostNode<TActor>(
        this.db,
        fbb,
        this.authContext,
      );
    }

    *[Symbol.iterator](): Generator<PostNode<TActor>> {
      for (const [id, buffer] of this.db.scan(this.collectionName)) {
        const fbb = PostFB.Post.getRootAsPost(new ByteBuffer(buffer));
        yield new PostNode<TActor>(
          this.db,
          fbb,
          this.authContext,
        );
      }
    }
  }

export class ZgClient<TActor> extends ZgDatabase {
  public readonly users: UserCollection<TActor>;
  public readonly posts: PostCollection<TActor>;
  constructor(
    prollyTree: ProllyTree,
    config: ZgConfig<TActor> = { entityResolvers: {}, globalResolvers: {} },
    authContext: ZgAuthContext<TActor> | null = null,
  ) {
    super(prollyTree, config, authContext);
    this.users = new UserCollection<TActor>(this, this.authContext);
    this.posts = new PostCollection<TActor>(this, this.authContext);
  }

  public with<TActor>(actor: TActor): ZgClient<TActor> {
    return new ZgClient<TActor>(this.prollyTree, this.config, { actor });
  }
}"
`;
